{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 150,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "import re\n",
    "import itertools\n",
    "import synapseclient\n",
    "from synapseclient import Activity"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 137,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Welcome, James Eddy!\n",
      "\n"
     ]
    }
   ],
   "source": [
    "syn = synapseclient.login()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## FASTQs, BAMs, counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def get_star_inputs(syn, star_input_folder_id):\n",
    "    \"\"\"\n",
    "    Collect full list of input BAM Synapse entities for the\n",
    "    current study (most of this data in the AMP-AD knowledge\n",
    "    portal is not annotated, so can't directly filter with\n",
    "    query). Note: inputs can also include FASTQs with\n",
    "    unmapped reads.\n",
    "    \"\"\"\n",
    "    query = ('select id, name from file where parentId==\"{}\"' \n",
    "             .format(star_input_folder_id))\n",
    "    return list(syn.chunkedQuery(query))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def get_star_out_r1_fastqs(syn, star_out_fastq_folder_id):\n",
    "    \"\"\"\n",
    "    Collect the R1 FASTQ Synapse entities generated by the STAR\n",
    "    reprocessing activity for the current study.\n",
    "    \"\"\"\n",
    "    fastq_query = ('select id, name from file where parentId==\"{}\" '\n",
    "                   'and read==\"R1\"'\n",
    "                   .format(star_out_fastq_folder_id))\n",
    "    return syn.chunkedQuery(fastq_query)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def get_sample_star_inputs(input_list, sample_id):\n",
    "    \"\"\"\n",
    "    Collect Synapse IDs of any input files matching the sample ID.\n",
    "    \"\"\"\n",
    "    return [entity['file.id'] for entity in input_list\n",
    "            if re.search(sample_id, entity['file.name'])]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def get_sample_r2_fastq(syn, star_out_fastq_folder_id, sample_id):\n",
    "    \"\"\"\n",
    "    Collect Synapse ID for R2 FASTQ generated by STAR and\n",
    "    matching the sample ID.\n",
    "    \"\"\"\n",
    "    fastq_query = ('select id, name from file where parentId==\"{}\" '\n",
    "                    'and specimenID==\"{}\" and read==\"R2\"'\n",
    "                    .format(star_out_fastq_folder_id, sample_id))\n",
    "    try:\n",
    "        return syn.chunkedQuery(fastq_query).next()\n",
    "    except StopIteration:\n",
    "        return"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def get_sample_counts(syn, star_out_count_folder_id, sample_id):\n",
    "    \"\"\"\n",
    "    Collect Synapse ID for counts generated by STAR and\n",
    "    matching the sample ID.\n",
    "    \"\"\"\n",
    "    counts_query = ('select id, name from file where parentId==\"{}\" '\n",
    "                    'and specimenID==\"{}\"'\n",
    "                    .format(star_out_counts_folder_id, sample_id))\n",
    "    return syn.chunkedQuery(counts_query).next()['file.id']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def get_sample_bam(syn, star_out_bam_folder_id, sample_id):\n",
    "    \"\"\"\n",
    "    Collect Synapse ID for BAM file generated by STAR and\n",
    "    matching the sample ID.\n",
    "    \"\"\"\n",
    "    bam_query = ('select id, name from file where parentId==\"{}\" '\n",
    "                    'and specimenID==\"{}\"'\n",
    "                    .format(star_out_bam_folder_id, sample_id))\n",
    "    return syn.chunkedQuery(bam_query).next()['file.id']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 158,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def set_star_output_provenance(syn, star_out_fastq_folder_id, \n",
    "                               star_input_list, star_script_id):\n",
    "    # get R1 FASTQs on Synapse for all study samples\n",
    "    star_out_r1_fastqs = get_star_out_r1_fastqs(syn, out_fastq_folder_id)\n",
    "    \n",
    "    for sample_r1_fastq in itertools.islice(star_out_r1_fastqs, 0, 2):\n",
    "        # parse sample ID from R1 FASTQ filename\n",
    "        sample_id = sample_r1_fastq['file.name'].split('.')[0]\n",
    "        print(\"Collecting inputs/outputs for R1 FASTQ ID: '{}' (sample ID: '{}')\"\n",
    "              .format(sample_r1_fastq['file.id'], sample_id))\n",
    "\n",
    "        # retrieve Synapse IDs for matched BAM input files\n",
    "        # based on the current sample ID\n",
    "        print(\"=> collecting matched input files from '{}'\".format(star_input_folder_id))\n",
    "        sample_star_input_ids = get_sample_star_inputs(input_list, sample_id)\n",
    "\n",
    "        # build activity for current sample inputs\n",
    "        print(\"=> building activity for script '{}'\".format(star_script_id))\n",
    "        sample_star_activity = Activity(\n",
    "            used=sample_star_input_ids, \n",
    "            executed=star_script_id\n",
    "        )\n",
    "        print(\"\\tActivity: used={}, executed={}\".format(\n",
    "              [u['reference']['targetId'] for u in sample_star_activity['used']\n",
    "               if not u['wasExecuted']],\n",
    "              [u['reference']['targetId'] for u in sample_star_activity['used']\n",
    "               if u['wasExecuted']])\n",
    "             )\n",
    "\n",
    "        # retrieve Synapse IDs for matched R2 FASTQ, reprocessed BAM,\n",
    "        # and counts file based on current sample ID\n",
    "        print(\"=> getting matched R2 FASTQ from '{}'\".format(star_out_fastq_folder_id))\n",
    "        sample_r2_fastq = get_sample_r2_fastq(syn, star_out_fastq_folder_id, sample_id)\n",
    "\n",
    "        print(\"=> getting matched counts from '{}'\".format(star_out_counts_folder_id))\n",
    "        sample_counts = get_sample_counts(syn, star_out_counts_folder_id, sample_id)\n",
    "\n",
    "        print(\"=> getting matched BAM from '{}'\".format(star_out_bam_folder_id))\n",
    "        sample_bam = get_sample_bam(syn, star_out_bam_folder_id, sample_id)\n",
    "\n",
    "        # combine sample output IDs\n",
    "        if sample_r2_fastq:\n",
    "            sample_star_output_ids = [\n",
    "                sample_r1_fastq['file.id'], \n",
    "                sample_r2_fastq['file.id'], \n",
    "                sample_counts, \n",
    "                sample_bam,\n",
    "            ]\n",
    "        else:\n",
    "            sample_star_output_ids = [\n",
    "                sample_r1_fastq['file.id'], \n",
    "                sample_counts, \n",
    "                sample_bam,\n",
    "            ]\n",
    "\n",
    "        # set provenance for all sample outputs\n",
    "        print(\"=> setting provenance for the following outputs\")\n",
    "        for output_id in sample_star_output_ids:\n",
    "            print(\"\\tOutput ID: '{}'\".format(output_id))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### MSSM"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# site-contributed Synapse BAMs and unmapped FASTQs for MSSM\n",
    "star_input_folder_id = 'syn7416949'\n",
    "star_input_list = get_star_inputs(syn, star_input_folder_id)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 159,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collecting inputs/outputs for R1 FASTQ ID: 'syn8613453' (sample ID: 'hB_RNA_12838')\n",
      "=> collecting matched input files from 'syn7416949'\n",
      "=> building activity for script 'run_star_mssm.sh'\n",
      "\tActivity: used=[u'syn5518922', u'syn5519399'], executed=['run_star_mssm.sh']\n",
      "=> getting matched R2 FASTQ from 'syn8540822'\n",
      "=> getting matched counts from 'syn8672879'\n",
      "=> getting matched BAM from 'syn8540822'\n",
      "=> setting provenance for the following outputs\n",
      "\tOutput ID: 'syn8613453'\n",
      "\tOutput ID: 'syn8686958'\n",
      "\tOutput ID: 'syn8596856'\n",
      "Collecting inputs/outputs for R1 FASTQ ID: 'syn8613635' (sample ID: 'BM_22_154')\n",
      "=> collecting matched input files from 'syn7416949'\n",
      "=> building activity for script 'run_star_mssm.sh'\n",
      "\tActivity: used=[u'syn4055855', u'syn4055856'], executed=['run_star_mssm.sh']\n",
      "=> getting matched R2 FASTQ from 'syn8540822'\n",
      "=> getting matched counts from 'syn8672879'\n",
      "=> getting matched BAM from 'syn8540822'\n",
      "=> setting provenance for the following outputs\n",
      "\tOutput ID: 'syn8613635'\n",
      "\tOutput ID: 'syn8686697'\n",
      "\tOutput ID: 'syn8593003'\n"
     ]
    }
   ],
   "source": [
    "# reprocessing output folders on Synapse for MSSM study\n",
    "star_out_fastq_folder_id = 'syn8612191'\n",
    "star_out_counts_folder_id = 'syn8672879'\n",
    "star_out_bam_folder_id = 'syn8540822'\n",
    "\n",
    "# Synapse ID for STAR reprocessing script for MSSM samples\n",
    "star_script_id = 'run_star_mssm.sh'\n",
    "\n",
    "# do things\n",
    "set_star_output_provenance(syn, star_out_bam_folder_id, star_input_list, star_script_id)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### ROSMAP"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# site-contributed Synapse BAMs for ROSMAP\n",
    "star_input_folder_id = 'syn4164376'\n",
    "star_input_list = get_star_inputs(syn, star_input_folder_id)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# reprocessing output folders on Synapse for ROSMAP study\n",
    "star_out_fastq_folder_id = 'syn8612097'\n",
    "star_out_counts_folder_id = 'syn8683042'\n",
    "star_out_bam_folder_id = 'syn8540863'\n",
    "\n",
    "# Synapse ID for STAR reprocessing script for ROSMAP samples\n",
    "star_script_id = 'run_star_rosmap.sh'\n",
    "\n",
    "set_star_output_provenance(syn, star_out_bam_folder_id, star_input_list, star_script_id)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Mayo_TCX"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# site-contributed Synapse BAMs for Mayo_TCX\n",
    "star_input_folder_id = 'syn4894912'\n",
    "star_input_list = get_star_inputs(syn, star_input_folder_id)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# reprocessing output folders on Synapse for Mayo_TCX study\n",
    "star_out_fastq_folder_id = 'syn8612203'\n",
    "star_out_counts_folder_id = 'syn8672882'\n",
    "star_out_bam_folder_id = 'syn8540820'\n",
    "\n",
    "# Synapse ID for STAR reprocessing script for Mayo_TCX samples\n",
    "star_script_id = 'run_star_mayo.sh'\n",
    "\n",
    "set_star_output_provenance(syn, star_out_bam_folder_id, star_input_list, star_script_id)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Mayo_CBE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# site-contributed Synapse BAMs for Mayo_CBE\n",
    "star_input_folder_id = 'syn5049322'\n",
    "star_input_list = get_star_inputs(syn, star_input_folder_id)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# reprocessing output folders on Synapse for Mayo_CBE study\n",
    "star_out_fastq_folder_id = 'syn8612213'\n",
    "star_out_counts_folder_id = 'syn8672881'\n",
    "star_out_bam_folder_id = 'syn8540821'\n",
    "\n",
    "# Synapse ID for STAR reprocessing script for Mayo_CBE samples\n",
    "star_script_id = 'run_star_mayo.sh'\n",
    "\n",
    "set_star_output_provenance(syn, star_out_bam_folder_id, star_input_list, star_script_id)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Mayo (failed QC)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "'syn9688855'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Picard alignment metrics"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "picard_out_folder_id = 'syn8683000' # ROSMAP\n",
    "picard_out_folder_id = 'syn8673913' # MSSM\n",
    "picard_out_folder_id = 'syn8673915' # Mayo_TCX\n",
    "picard_out_folder_id = 'syn8663914' # Mayo_CBE"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Combined counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "study_counts_id = 'syn8691134.1' # ROSMAP\n",
    "study_counts_id = 'syn8691099.1' # MSSM\n",
    "study_counts_id = 'syn8690799.1' # Mayo_TCX\n",
    "study_counts_id = 'syn8690904.1' # Mayo_CBE"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Combined metrics"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "study_counts_id = 'syn8698240.2' # ROSMAP\n",
    "study_counts_id = 'syn8698270.1' # MSSM\n",
    "study_counts_id = 'syn8698211.1' # Mayo_TCX\n",
    "study_counts_id = 'syn8698214.1' # Mayo_CBE"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
