{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import sys\n",
    "import re\n",
    "import itertools\n",
    "import synapseclient\n",
    "from synapseclient import Activity"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Welcome, James Eddy!\n",
      "\n"
     ]
    }
   ],
   "source": [
    "syn = synapseclient.login()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## FASTQs, BAMs, counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def get_star_inputs(syn, star_input_folder_id):\n",
    "    \"\"\"\n",
    "    Collect full list of input BAM Synapse entities for the\n",
    "    current study (most of this data in the AMP-AD knowledge\n",
    "    portal is not annotated, so can't directly filter with\n",
    "    query). Note: inputs can also include FASTQs with\n",
    "    unmapped reads.\n",
    "    \"\"\"\n",
    "    query = ('select id, name from file where parentId==\"{}\"' \n",
    "             .format(star_input_folder_id))\n",
    "    return list(syn.chunkedQuery(query))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def get_star_out_r1_fastqs(syn, star_out_fastq_folder_id):\n",
    "    \"\"\"\n",
    "    Collect the R1 FASTQ Synapse entities generated by the STAR\n",
    "    reprocessing activity for the current study.\n",
    "    \"\"\"\n",
    "    fastq_query = ('select id, name from file where parentId==\"{}\" '\n",
    "                   'and read==\"R1\"'\n",
    "                   .format(star_out_fastq_folder_id))\n",
    "    return syn.chunkedQuery(fastq_query)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def get_sample_star_inputs(input_list, sample_id):\n",
    "    \"\"\"\n",
    "    Collect Synapse IDs of any input files matching the sample ID.\n",
    "    \"\"\"\n",
    "    return [entity['file.id'] for entity in input_list\n",
    "            if re.search(sample_id, entity['file.name'])]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def get_sample_r2_fastq(syn, star_out_fastq_folder_id, sample_id):\n",
    "    \"\"\"\n",
    "    Collect Synapse ID for R2 FASTQ generated by STAR and\n",
    "    matching the sample ID.\n",
    "    \"\"\"\n",
    "    fastq_query = ('select id, name from file where parentId==\"{}\" '\n",
    "                    'and specimenID==\"{}\" and read==\"R2\"'\n",
    "                    .format(star_out_fastq_folder_id, sample_id))\n",
    "    try:\n",
    "        return syn.chunkedQuery(fastq_query).next()\n",
    "    except StopIteration:\n",
    "        return"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def get_sample_counts(syn, star_out_count_folder_id, sample_id):\n",
    "    \"\"\"\n",
    "    Collect Synapse ID for counts generated by STAR and\n",
    "    matching the sample ID.\n",
    "    \"\"\"\n",
    "    counts_query = ('select id, name from file where parentId==\"{}\" '\n",
    "                    'and specimenID==\"{}\"'\n",
    "                    .format(star_out_counts_folder_id, sample_id))\n",
    "    return syn.chunkedQuery(counts_query).next()['file.id']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def get_sample_bam(syn, star_out_bam_folder_id, sample_id):\n",
    "    \"\"\"\n",
    "    Collect Synapse ID for BAM file generated by STAR and\n",
    "    matching the sample ID.\n",
    "    \"\"\"\n",
    "    bam_query = ('select id, name from file where parentId==\"{}\" '\n",
    "                    'and specimenID==\"{}\"'\n",
    "                    .format(star_out_bam_folder_id, sample_id))\n",
    "    return syn.chunkedQuery(bam_query).next()['file.id']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def set_star_output_provenance(syn, star_out_fastq_folder_id, \n",
    "                               star_input_list, star_script_id):\n",
    "    # get R1 FASTQs on Synapse for all study samples\n",
    "    star_out_r1_fastqs = get_star_out_r1_fastqs(syn, star_out_fastq_folder_id)\n",
    "    \n",
    "    for sample_r1_fastq in itertools.islice(star_out_r1_fastqs, 0, 2):\n",
    "        # parse sample ID from R1 FASTQ filename\n",
    "        sample_id = sample_r1_fastq['file.name'].split('.')[0]\n",
    "        print(\"Collecting inputs/outputs for R1 FASTQ ID: '{}' (sample ID: '{}')\"\n",
    "              .format(sample_r1_fastq['file.id'], sample_id))\n",
    "\n",
    "        # retrieve Synapse IDs for matched BAM input files\n",
    "        # based on the current sample ID\n",
    "        print(\"=> collecting matched input files from '{}'\".format(star_input_folder_id))\n",
    "        sample_star_input_ids = get_sample_star_inputs(star_input_list, sample_id)\n",
    "\n",
    "        # build activity for current sample inputs\n",
    "        print(\"=> building activity for script '{}'\".format(star_script_id))\n",
    "        sample_star_activity = Activity(\n",
    "            used=sample_star_input_ids, \n",
    "            executed=star_script_id\n",
    "        )\n",
    "        print(\"\\tActivity: used={}, executed={}\".format(\n",
    "              [u['reference']['targetId'] for u in sample_star_activity['used']\n",
    "               if not u['wasExecuted']],\n",
    "              [u['reference']['targetId'] for u in sample_star_activity['used']\n",
    "               if u['wasExecuted']])\n",
    "             )\n",
    "\n",
    "        # retrieve Synapse IDs for matched R2 FASTQ, reprocessed BAM,\n",
    "        # and counts file based on current sample ID\n",
    "        print(\"=> getting matched R2 FASTQ from '{}'\".format(star_out_fastq_folder_id))\n",
    "        sample_r2_fastq = get_sample_r2_fastq(syn, star_out_fastq_folder_id, sample_id)\n",
    "\n",
    "        print(\"=> getting matched counts from '{}'\".format(star_out_counts_folder_id))\n",
    "        sample_counts = get_sample_counts(syn, star_out_counts_folder_id, sample_id)\n",
    "\n",
    "        print(\"=> getting matched BAM from '{}'\".format(star_out_bam_folder_id))\n",
    "        sample_bam = get_sample_bam(syn, star_out_bam_folder_id, sample_id)\n",
    "\n",
    "        # combine sample output IDs\n",
    "        if sample_r2_fastq:\n",
    "            sample_star_output_ids = [\n",
    "                sample_r1_fastq['file.id'], \n",
    "                sample_r2_fastq['file.id'], \n",
    "                sample_counts, \n",
    "                sample_bam,\n",
    "            ]\n",
    "        else:\n",
    "            sample_star_output_ids = [\n",
    "                sample_r1_fastq['file.id'], \n",
    "                sample_counts, \n",
    "                sample_bam,\n",
    "            ]\n",
    "\n",
    "        # set provenance for all sample outputs\n",
    "        print(\"=> setting provenance for the following outputs\")\n",
    "        for output_id in sample_star_output_ids:\n",
    "            print(\"\\tOutput ID: '{}'\".format(output_id))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### MSSM"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# site-contributed Synapse BAMs and unmapped FASTQs for MSSM\n",
    "star_input_folder_id = 'syn7416949'\n",
    "star_input_list = get_star_inputs(syn, star_input_folder_id)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collecting inputs/outputs for R1 FASTQ ID: 'syn8620246' (sample ID: 'BM_10_546')\n",
      "=> collecting matched input files from 'syn7416949'\n",
      "=> building activity for script 'run_star_mssm.sh'\n",
      "\tActivity: used=[u'syn4055333', u'syn4055334'], executed=['run_star_mssm.sh']\n",
      "=> getting matched R2 FASTQ from 'syn8612191'\n",
      "=> getting matched counts from 'syn8672879'\n",
      "=> getting matched BAM from 'syn8540822'\n",
      "=> setting provenance for the following outputs\n",
      "\tOutput ID: 'syn8620246'\n",
      "\tOutput ID: 'syn8686802'\n",
      "\tOutput ID: 'syn8593437'\n",
      "Collecting inputs/outputs for R1 FASTQ ID: 'syn8669905' (sample ID: 'BM_10_548')\n",
      "=> collecting matched input files from 'syn7416949'\n",
      "=> building activity for script 'run_star_mssm.sh'\n",
      "\tActivity: used=[u'syn4055335', u'syn4055336'], executed=['run_star_mssm.sh']\n",
      "=> getting matched R2 FASTQ from 'syn8612191'\n",
      "=> getting matched counts from 'syn8672879'\n",
      "=> getting matched BAM from 'syn8540822'\n",
      "=> setting provenance for the following outputs\n",
      "\tOutput ID: 'syn8669905'\n",
      "\tOutput ID: 'syn8687585'\n",
      "\tOutput ID: 'syn8604616'\n"
     ]
    }
   ],
   "source": [
    "# reprocessing output folders on Synapse for MSSM study\n",
    "star_out_fastq_folder_id = 'syn8612191'\n",
    "star_out_counts_folder_id = 'syn8672879'\n",
    "star_out_bam_folder_id = 'syn8540822'\n",
    "\n",
    "# Synapse ID for STAR reprocessing script for MSSM samples\n",
    "star_script_id = 'run_star_mssm.sh'\n",
    "\n",
    "# do things\n",
    "set_star_output_provenance(syn, star_out_fastq_folder_id, star_input_list, star_script_id)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### ROSMAP"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# site-contributed Synapse BAMs for ROSMAP\n",
    "star_input_folder_id = 'syn4164376'\n",
    "star_input_list = get_star_inputs(syn, star_input_folder_id)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collecting inputs/outputs for R1 FASTQ ID: 'syn8613921' (sample ID: '108_120418')\n",
      "=> collecting matched input files from 'syn4164376'\n",
      "=> building activity for script 'run_star_rosmap.sh'\n",
      "\tActivity: used=[u'syn4212548'], executed=['run_star_rosmap.sh']\n",
      "=> getting matched R2 FASTQ from 'syn8612097'\n",
      "=> getting matched counts from 'syn8683042'\n"
     ]
    },
    {
     "ename": "StopIteration",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mStopIteration\u001b[0m                             Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-28-9e63ed9c9ef1>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      7\u001b[0m \u001b[0mstar_script_id\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m'run_star_rosmap.sh'\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      8\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 9\u001b[0;31m \u001b[0mset_star_output_provenance\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0msyn\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mstar_out_fastq_folder_id\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mstar_input_list\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mstar_script_id\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m<ipython-input-23-d488686f30df>\u001b[0m in \u001b[0;36mset_star_output_provenance\u001b[0;34m(syn, star_out_fastq_folder_id, star_input_list, star_script_id)\u001b[0m\n\u001b[1;32m     34\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     35\u001b[0m         \u001b[0;32mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"=> getting matched counts from '{}'\"\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mformat\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mstar_out_counts_folder_id\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 36\u001b[0;31m         \u001b[0msample_counts\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mget_sample_counts\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0msyn\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mstar_out_counts_folder_id\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0msample_id\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     37\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     38\u001b[0m         \u001b[0;32mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"=> getting matched BAM from '{}'\"\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mformat\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mstar_out_bam_folder_id\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m<ipython-input-7-0419bca12804>\u001b[0m in \u001b[0;36mget_sample_counts\u001b[0;34m(syn, star_out_count_folder_id, sample_id)\u001b[0m\n\u001b[1;32m      7\u001b[0m                     \u001b[0;34m'and specimenID==\"{}\"'\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      8\u001b[0m                     .format(star_out_counts_folder_id, sample_id))\n\u001b[0;32m----> 9\u001b[0;31m     \u001b[0;32mreturn\u001b[0m \u001b[0msyn\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mchunkedQuery\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcounts_query\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mnext\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'file.id'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;31mStopIteration\u001b[0m: "
     ]
    }
   ],
   "source": [
    "# reprocessing output folders on Synapse for ROSMAP study\n",
    "star_out_fastq_folder_id = 'syn8612097'\n",
    "star_out_counts_folder_id = 'syn8683042'\n",
    "star_out_bam_folder_id = 'syn8540863'\n",
    "\n",
    "# Synapse ID for STAR reprocessing script for ROSMAP samples\n",
    "star_script_id = 'run_star_rosmap.sh'\n",
    "\n",
    "set_star_output_provenance(syn, star_out_fastq_folder_id, star_input_list, star_script_id)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Mayo_TCX"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# site-contributed Synapse BAMs for Mayo_TCX\n",
    "star_input_folder_id = 'syn4894912'\n",
    "star_input_list = get_star_inputs(syn, star_input_folder_id)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collecting inputs/outputs for R1 FASTQ ID: 'syn8619777' (sample ID: '1005_TCX')\n",
      "=> collecting matched input files from 'syn4894912'\n",
      "=> building activity for script 'run_star_mayo.sh'\n",
      "\tActivity: used=[u'syn4898414'], executed=['run_star_mayo.sh']\n",
      "=> getting matched R2 FASTQ from 'syn8612203'\n",
      "=> getting matched counts from 'syn8672882'\n",
      "=> getting matched BAM from 'syn8540820'\n",
      "=> setting provenance for the following outputs\n",
      "\tOutput ID: 'syn8619777'\n",
      "\tOutput ID: 'syn8621132'\n",
      "\tOutput ID: 'syn8685924'\n",
      "\tOutput ID: 'syn8620557'\n",
      "Collecting inputs/outputs for R1 FASTQ ID: 'syn8619364' (sample ID: '1010_TCX')\n",
      "=> collecting matched input files from 'syn4894912'\n",
      "=> building activity for script 'run_star_mayo.sh'\n",
      "\tActivity: used=[u'syn4899441'], executed=['run_star_mayo.sh']\n",
      "=> getting matched R2 FASTQ from 'syn8612203'\n",
      "=> getting matched counts from 'syn8672882'\n",
      "=> getting matched BAM from 'syn8540820'\n",
      "=> setting provenance for the following outputs\n",
      "\tOutput ID: 'syn8619364'\n",
      "\tOutput ID: 'syn8620580'\n",
      "\tOutput ID: 'syn8686121'\n",
      "\tOutput ID: 'syn8570047'\n"
     ]
    }
   ],
   "source": [
    "# reprocessing output folders on Synapse for Mayo_TCX study\n",
    "star_out_fastq_folder_id = 'syn8612203'\n",
    "star_out_counts_folder_id = 'syn8672882'\n",
    "star_out_bam_folder_id = 'syn8540820'\n",
    "\n",
    "# Synapse ID for STAR reprocessing script for Mayo_TCX samples\n",
    "star_script_id = 'run_star_mayo.sh'\n",
    "\n",
    "set_star_output_provenance(syn, star_out_fastq_folder_id, star_input_list, star_script_id)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Mayo_CBE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# site-contributed Synapse BAMs for Mayo_CBE\n",
    "star_input_folder_id = 'syn5049322'\n",
    "star_input_list = get_star_inputs(syn, star_input_folder_id)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[{u'file.id': u'syn5196750',\n",
       "  u'file.name': u'1000_CER.FCC7KUNACXX_L5IATCACG.snap.bam'},\n",
       " {u'file.id': u'syn5051379',\n",
       "  u'file.name': u'1010_CER.FCC7DJAACXX_L5ITTAGGC.snap.bam'},\n",
       " {u'file.id': u'syn5051388',\n",
       "  u'file.name': u'1015_CER.FCC7983ACXX_L5IACTTGA.snap.bam'},\n",
       " {u'file.id': u'syn5051449',\n",
       "  u'file.name': u'1019_CER.FCC7BR1ACXX_L7IGTGGCC.snap.bam'},\n",
       " {u'file.id': u'syn5051608',\n",
       "  u'file.name': u'1027_CER.FCC7BR1ACXX_L2IGATCAG.snap.bam'}]"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "star_input_list[0:5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# reprocessing output folders on Synapse for Mayo_CBE study\n",
    "star_out_fastq_folder_id = 'syn8612213'\n",
    "star_out_counts_folder_id = 'syn8672881'\n",
    "star_out_bam_folder_id = 'syn8540821'\n",
    "\n",
    "# Synapse ID for STAR reprocessing script for Mayo_CBE samples\n",
    "star_script_id = 'run_star_mayo.sh'\n",
    "\n",
    "set_star_output_provenance(syn, star_out_fastq_folder_id, star_input_list, star_script_id)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Mayo (failed QC)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "'syn9688855'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Picard alignment metrics"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def get_picard_out_metrics(syn, picard_out_metrics_folder_id):\n",
    "    \"\"\"\n",
    "    Collect the metrics table Synapse entities generated by the\n",
    "    Picard activity for the current study.\n",
    "    \"\"\"\n",
    "    metrics_query = ('select id, name from file where parentId==\"{}\"'\n",
    "                     .format(picard_out_metrics_folder_id))\n",
    "    return syn.chunkedQuery(metrics_query)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def get_sample_picard_input(syn, picard_input_folder_id, sample_id):\n",
    "    \"\"\"\n",
    "    Collect Synapse IDs for input BAM file matching the sample ID.\n",
    "    \"\"\"\n",
    "    bam_query = ('select id, name from file where parentId==\"{}\" '\n",
    "                 'and specimenID==\"{}\"'\n",
    "                 .format(picard_input_folder_id, sample_id))\n",
    "    return syn.chunkedQuery(bam_query).next()['file.id']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def set_picard_output_provenance(syn, picard_out_folder_id, \n",
    "                                 picard_input_folder_id, picard_script_id):\n",
    "    # get metrics on Synapse for all study samples\n",
    "    picard_out_metrics = get_picard_out_metrics(syn, picard_out_folder_id)\n",
    "    \n",
    "    for sample_metrics in itertools.islice(picard_out_metrics, 0, 2):\n",
    "        # parse sample ID from metrics filename\n",
    "        sample_id = sample_metrics['file.name'].split('.')[0].strip('_picard')\n",
    "        print(\"Collecting inputs/outputs for metrics ID: '{}' (sample ID: '{}')\"\n",
    "              .format(sample_metrics['file.id'], sample_id))\n",
    "\n",
    "        # retrieve Synapse ID for matched input BAM file\n",
    "        print(\"=> getting matched BAM from '{}'\".format(picard_input_folder_id))\n",
    "        sample_bam = get_sample_picard_input(syn, picard_input_folder_id, sample_id)\n",
    "\n",
    "        # build activity for current sample inputs\n",
    "        print(\"=> building activity for script '{}'\".format(picard_script_id))\n",
    "        sample_picard_activity = Activity(\n",
    "            used=sample_bam, \n",
    "            executed=picard_script_id\n",
    "        )\n",
    "        print(\"\\tActivity: used={}, executed={}\".format(\n",
    "              [u['reference']['targetId'] for u in sample_picard_activity['used']\n",
    "               if not u['wasExecuted']],\n",
    "              [u['reference']['targetId'] for u in sample_picard_activity['used']\n",
    "               if u['wasExecuted']])\n",
    "             )\n",
    "\n",
    "        # set provenance for all sample outputs\n",
    "        print(\"=> setting provenance for the following outputs\")\n",
    "        print(\"\\tOutput ID: '{}'\".format(sample_metrics['file.id']))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collecting inputs/outputs for metrics ID: 'syn8683105' (sample ID: 'BM_10_546')\n",
      "=> getting matched BAM from 'syn8540822'\n",
      "=> building activity for script 'run_picard.sh'\n",
      "\tActivity: used=[u'syn8593437'], executed=['run_picard.sh']\n",
      "=> setting provenance for the following outputs\n",
      "\tOutput ID: 'syn8683105'\n",
      "Collecting inputs/outputs for metrics ID: 'syn8683891' (sample ID: 'BM_10_548')\n",
      "=> getting matched BAM from 'syn8540822'\n",
      "=> building activity for script 'run_picard.sh'\n",
      "\tActivity: used=[u'syn8604616'], executed=['run_picard.sh']\n",
      "=> setting provenance for the following outputs\n",
      "\tOutput ID: 'syn8683891'\n"
     ]
    }
   ],
   "source": [
    "picard_out_folder_id = 'syn8673913' # MSSM\n",
    "picard_input_folder_id = 'syn8540822'\n",
    "picard_script_id = 'run_picard.sh'\n",
    "\n",
    "set_picard_output_provenance(syn, picard_out_folder_id, \n",
    "                             picard_input_folder_id, picard_script_id)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collecting inputs/outputs for metrics ID: 'syn8685374' (sample ID: '01_120405')\n",
      "=> getting matched BAM from 'syn8540863'\n"
     ]
    },
    {
     "ename": "StopIteration",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mStopIteration\u001b[0m                             Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-47-6d2f0134c589>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      4\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      5\u001b[0m set_picard_output_provenance(syn, picard_out_folder_id, \n\u001b[0;32m----> 6\u001b[0;31m                              picard_input_folder_id, picard_script_id)\n\u001b[0m",
      "\u001b[0;32m<ipython-input-45-fc886d3c1c08>\u001b[0m in \u001b[0;36mset_picard_output_provenance\u001b[0;34m(syn, picard_out_folder_id, picard_input_folder_id, picard_script_id)\u001b[0m\n\u001b[1;32m     12\u001b[0m         \u001b[0;31m# retrieve Synapse ID for matched input BAM file\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     13\u001b[0m         \u001b[0;32mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"=> getting matched BAM from '{}'\"\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mformat\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mpicard_input_folder_id\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 14\u001b[0;31m         \u001b[0msample_bam\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mget_sample_picard_input\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0msyn\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mpicard_input_folder_id\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0msample_id\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     15\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     16\u001b[0m         \u001b[0;31m# build activity for current sample inputs\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m<ipython-input-41-d20bf9ebda49>\u001b[0m in \u001b[0;36mget_sample_picard_input\u001b[0;34m(syn, picard_input_folder_id, sample_id)\u001b[0m\n\u001b[1;32m      6\u001b[0m                     \u001b[0;34m'and specimenID==\"{}\"'\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      7\u001b[0m                     .format(picard_input_folder_id, sample_id))\n\u001b[0;32m----> 8\u001b[0;31m     \u001b[0;32mreturn\u001b[0m \u001b[0msyn\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mchunkedQuery\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mbam_query\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mnext\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'file.id'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;31mStopIteration\u001b[0m: "
     ]
    }
   ],
   "source": [
    "picard_out_folder_id = 'syn8683000' # ROSMAP\n",
    "picard_input_folder_id = 'syn8540863'\n",
    "picard_script_id = 'run_picard.sh'\n",
    "\n",
    "set_picard_output_provenance(syn, picard_out_folder_id, \n",
    "                             picard_input_folder_id, picard_script_id)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collecting inputs/outputs for metrics ID: 'syn8683162' (sample ID: '1005_TCX')\n",
      "=> getting matched BAM from 'syn8540820'\n",
      "=> building activity for script 'run_picard.sh'\n",
      "\tActivity: used=[u'syn8620557'], executed=['run_picard.sh']\n",
      "=> setting provenance for the following outputs\n",
      "\tOutput ID: 'syn8683162'\n",
      "Collecting inputs/outputs for metrics ID: 'syn8682855' (sample ID: '1010_TCX')\n",
      "=> getting matched BAM from 'syn8540820'\n",
      "=> building activity for script 'run_picard.sh'\n",
      "\tActivity: used=[u'syn8570047'], executed=['run_picard.sh']\n",
      "=> setting provenance for the following outputs\n",
      "\tOutput ID: 'syn8682855'\n"
     ]
    }
   ],
   "source": [
    "picard_out_folder_id = 'syn8673915' # Mayo_TCX\n",
    "picard_input_folder_id = 'syn8540820'\n",
    "picard_script_id = 'run_picard.sh'\n",
    "\n",
    "set_picard_output_provenance(syn, picard_out_folder_id, \n",
    "                             picard_input_folder_id, picard_script_id)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collecting inputs/outputs for metrics ID: 'syn8684675' (sample ID: '1000_CER')\n",
      "=> getting matched BAM from 'syn8540821'\n",
      "=> building activity for script 'run_picard.sh'\n",
      "\tActivity: used=[u'syn8545498'], executed=['run_picard.sh']\n",
      "=> setting provenance for the following outputs\n",
      "\tOutput ID: 'syn8684675'\n",
      "Collecting inputs/outputs for metrics ID: 'syn8684470' (sample ID: '1010_CER')\n",
      "=> getting matched BAM from 'syn8540821'\n",
      "=> building activity for script 'run_picard.sh'\n",
      "\tActivity: used=[u'syn8541791'], executed=['run_picard.sh']\n",
      "=> setting provenance for the following outputs\n",
      "\tOutput ID: 'syn8684470'\n"
     ]
    }
   ],
   "source": [
    "picard_out_folder_id = 'syn8673914' # Mayo_CBE\n",
    "picard_input_folder_id = 'syn8540821'\n",
    "picard_script_id = 'run_picard.sh'\n",
    "\n",
    "set_picard_output_provenance(syn, picard_out_folder_id, \n",
    "                             picard_input_folder_id, picard_script_id)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Combined counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "study_counts_id = 'syn8691134.1' # ROSMAP\n",
    "study_counts_id = 'syn8691099.1' # MSSM\n",
    "study_counts_id = 'syn8690799.1' # Mayo_TCX\n",
    "study_counts_id = 'syn8690904.1' # Mayo_CBE"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Combined metrics"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "study_counts_id = 'syn8698240.2' # ROSMAP\n",
    "study_counts_id = 'syn8698270.1' # MSSM\n",
    "study_counts_id = 'syn8698211.1' # Mayo_TCX\n",
    "study_counts_id = 'syn8698214.1' # Mayo_CBE"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
